navigation/navigation-compose-material/src/main/java/androidx/navigation/compose/material/SheetContentHost.kt

```diff
@@ -0,0 +1,130 @@
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.DisposableEffect
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.runtime.rememberCoroutineScope
+import androidx.compose.runtime.saveable.SaveableStateHolder
+import androidx.compose.runtime.snapshotFlow
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.unit.dp
+import androidx.navigation.NavBackStackEntry
+import androidx.navigation.compose.LocalOwnersProvider
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.currentCoroutineContext
+import kotlinx.coroutines.ensureActive
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.drop
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.launch
+
+/**
+ * Hosts a [BottomSheetNavigator.Destination]'s [NavBackStackEntry] and its
+ * [BottomSheetNavigator.Destination.content] and provides a [onSheetDismissed] callback. It also
+ * shows and hides the [ModalBottomSheetLayout] through the [sheetState] when the sheet content
+ * enters or leaves the composition.
+ *
+ * @param columnHost The [ColumnScope] the sheet content is hosted in, typically the instance
+ * that is provided by [ModalBottomSheetLayout]
+ * @param backStackEntry The [NavBackStackEntry] holding the [BottomSheetNavigator.Destination],
+ * or null if there is no [NavBackStackEntry]
+ * @param sheetState The [ModalBottomSheetState] used to observe and control the sheet visibility
+ * @param onSheetDismissed Callback when the sheet has been dismissed. Typically, you'll want to
+ * pop the back stack here.
+ */
+@OptIn(ExperimentalMaterialApi::class)
+@Composable
+internal fun SheetContentHost(
+    columnHost: ColumnScope,
+    backStackEntry: NavBackStackEntry?,
+    sheetState: ModalBottomSheetState,
+    saveableStateHolder: SaveableStateHolder,
+    onSheetShown: (entry: NavBackStackEntry) -> Unit,
+    onSheetDismissed: (entry: NavBackStackEntry) -> Unit
+) {
+    val scope = rememberCoroutineScope()
+    if (backStackEntry != null) {
+        LaunchedEffect(Unit) {
+            val sheetVisibility = snapshotFlow { sheetState.isVisible }
+            sheetVisibility
+                // We are only interested in changes in the sheet's visibility
+                .distinctUntilChanged()
+                // distinctUntilChanged emits the initial value which we don't need
+                .drop(1)
+                // We want to know when the sheet was visible but is not anymore
+                .filter { isVisible -> !isVisible }
+                // Finally, pop the back stack when the sheet has been hidden
+                .collect { onSheetDismissed(backStackEntry) }
+        }
+
+        // Whenever the composable associated with the latestEntry enters the composition, we
+        // want to show the sheet, and hide it when this composable leaves the composition
+        DisposableEffect(Unit) {
```

Comment:
How does this handle navigating from one bottom sheet destination to another? I would have thought you'd need a `DisposableEffect(backStackEntry` or something similar here? When navigating between bottom sheets, what is the expected behavior here? I would think the first one would want to hide completely, then the second one shows? I'm not sure how to write that logic though.

Reply:
Yeah you're right about that, the effect should be scoped to the entry. The sheet would hide and then a new one appear - I'm not suuuuper happy with that because if you think about nesting destinations in a sheet you'll most likely want to have a transition along the x-axis. But I think that's out of scope for this PR :D

Reply:
Well, navigating to a new bottom sheet destination is not about replacing the content within the bottom sheet, but hiding the existing bottom sheet destination and then showing a brand new bottom sheet. That's the styling we should shoot for.

Reply:
Yup. Would be cool to enable something else in the future but we should stick to "hide then show" for now (that's also what it is)